### PR TITLE
Crash Dumper: E2E tests using kill filter

### LINKED ISSUE #14249: Expand the area of ScopeTrackedObject usage
Expand the area of `ScopeTrackedObject` usage.

Currently the `ScopeTrackedObject` is used at the L7 stream level. The mechanism essentially registers that the dispatcher is doing some particular work.

When a Fatal Error occurs on a thread, we invoke several debug actions on the dispatcher that encountered the error, trying to also leverage the `ScopeTrackedObject` (what it was working on at the time) to give additional insight into the crash. 

With this mechanism as is, there are many parts of the pipeline (see image below) that if we "burst" (crashed) there we'd get no additional information. Furthermore, even when we do crash at the L7 level we have room to improve what we dump -- we don't dump some useful L4 information such as IP address IIRC. 

Stream Pipeline

Ideally, we should be able to crash along any point of the pipeline and get information from the earlier phase from the pipeline as well. This will likely mean extending what objects implement the `ScopedTrackedObject` interface, and probably changing how the dispatcher tracks `ScopeTrackedObject`s, perhaps as a stack vs as a single object.

Let me know if that makes sense, thanks.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: E2E test for crash dumper using the kill filter
Additional Description: The separate components where we have crash dumping support have individual unit tests, some with more mocks than others. E2E tests will build more confidence in the crash dumping pipeline.
Risk Level: Low (tests only)
Testing: Integration tests
Docs Changes: NA
Release Notes: NA
Platform Specific Features: Systems that use the crash handlers (windows doesn't currently)
Fixes #14249

### CODE REVIEW

File: test/extensions/filters/http/kill_request/crash_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#include "envoy/extensions/filters/http/kill_request/v3/kill_request.pb.h"
+
+#include "test/integration/http_protocol_integration.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
```
Comment (htuch) [inline test/extensions/filters/http/kill_request/crash_integration_test.cc:8]: Nit: please follow standard test extension namespace conventions. Should be qualified with extension and also put inside an inner anonymous namespace.

**Review by htuch — COMMENTED** [review]
LGTM modulo nit. Thanks for improving testing!

File: test/extensions/filters/http/kill_request/crash_integration_test.cc (hunk 1 above)
Comment (KBaichoo) [reply test/extensions/filters/http/kill_request/crash_integration_test.cc:21]: Whoops. Done.

**Comment by KBaichoo** [comment]
/retest

Network flake

**Review by htuch — APPROVED** [review]

**Comment by htuch** [comment]
@KBaichoo seems a legit CI failure, can you take a look?
/wait

**Comment by KBaichoo** [comment]
Fixed -- the test doesn't make sense to run under the compile time options mode since it requires Envoy to handle signals which that mode disables.

**Pull request merged**